### PR TITLE
add validation/log for the skipped route due to catchall route

### PR DESCRIPTION
**Pull request by shonecyx**
**Please provide a description of this PR:**

Add log for https://github.com/istio/istio/issues/46410

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Dual Stack
- [ ] Installation
- [ x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/route/route.go
Comment (sword-jin) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:392]: I think this %v will print a super huge string. `Http` is a big structure. Leave this to other developers.

**Review by sword-jin — APPROVED** [review]

File: pilot/pkg/networking/core/v1alpha3/route/route.go
Comment (shonecyx) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:392]: It won't print if there is no http match gets bypassed, if it prints, then the huge string is bypassed, it will help for the troubleshooting since there was a breaking change made.

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:392]: I don't think we want to print this big message. I am not sure we want to print anything at all; istiod logs are not intended for app developer persona. Maybe this is better suited as an analyzer or warning in the webhook

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:392]: I donot mind if you change to debug, and , maybe viable to add the check  in validation webhook

Comment (shonecyx) [reply pilot/pkg/networking/core/v1alpha3/route/route.go]: The background is that it took us some time to figure out why some VirtualService [routes](https://github.com/istio/istio/issues/46410#issue-1842913565) were working in Isito 1.10 but got 404 after upgrading to `1.17`.  There isn't any log in istiod indicates the problem and I'd believe others will encounter this issue during the upgrade.

I changed to print the skipped route number instead of route itself, PTAL.

Change to debug might not help much since normally we won't run istio in debug mode and validation webhook seems feasible but I think we still can only log a warming message for backward compatible. 
cc @howardjohn  @hzxuzhonghu

Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/route/route.go]: I can get your meaning, can you update validation as well

Comment (shonecyx) [reply pilot/pkg/networking/core/v1alpha3/route/route.go]: Added validation, PTAL

File: pkg/config/validation/validation.go
Comment (hzxuzhonghu) [inline pkg/config/validation/validation.go:2696]: can we resue that from pilot/pkg/networking/core/v1alpha3/route/route.go

Comment (hzxuzhonghu) [reply pkg/config/validation/validation.go:2696]: Oh, that one can be moved here as not used there

File: pkg/config/validation/validation.go
Comment (hzxuzhonghu) [inline pkg/config/validation/validation.go:2684]: Isn't that all `route.Match` are catch all, then the route is catch all?

File: pkg/config/validation/validation.go
Comment (shonecyx) [reply pkg/config/validation/validation.go:2696]: `isCatchAllMatch` is actually copied from ` pilot/pkg/networking/core/v1alpha3/route/route.go`, it's being used for test cases only and can't import it due to circular dependency. Maybe to put this method in some util so it can be shared by both, any suggestion where to put?

File: pkg/config/validation/validation.go
Comment (shonecyx) [reply pkg/config/validation/validation.go:2684]: I'm afraid not, if there is one match of the route is catch all match, then it's catch all route and the match below the catchall matched will be bypassed. FYI https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/route/route.go#L376-L388

I also updated some of existing testing specs like below one to make it backward compatible.

From
```
  - name: "send product to sample.foo"
    match:
    - uri:
        prefix: "/"
    - uri:
        exact: "/"  
      method:
        exact: GET
```
To
```
  - name: "send product to sample.foo"
    match:
    - uri:
        exact: "/"
    - uri:
        prefix: "/"
      method:
        exact: GET
```

File: pkg/config/validation/validation.go
Comment (hzxuzhonghu) [reply pkg/config/validation/validation.go:2696]: test case can be moved too

File: pkg/config/validation/validation.go
Comment (hzxuzhonghu) [reply pkg/config/validation/validation.go:2684]: you are right

File: pkg/config/validation/validation.go
Comment (shonecyx) [reply pkg/config/validation/validation.go:2696]: Removed `isCatchAllMatch` in `route.go` but still keep those test cases

**Review by hzxuzhonghu — APPROVED** [review]
Thanks

File: pilot/pkg/networking/core/v1alpha3/route/route.go
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:392]:
```suggestion
				log.Warnf("Skipping %d routes for %s due to the catch all route", len(vs.Http)-i, virtualService.Name)
```

File: pkg/config/validation/validation.go
Comment (howardjohn) [reply pkg/config/validation/validation.go:2684]: The two YAML configs you provided above function 100% identically. We should not warn in these cases.

There are two lists: HTTPRoutes, and HTTPMatchRequests.

Above we have 1 HTTPRoute with 2 HTTPMatchRequest.

Within an HTTPRoute, the order of HTTPMatchRequest is irrelevant. If any match, the request route matches; it doesn't matter _which one_ matches.

**Across** HTTPRoutes is where it matters

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: pilot/pkg/networking/core/v1alpha3/route/route.go
Comment (shonecyx) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:392]: updated

File: pkg/config/validation/validation.go
Comment (shonecyx) [reply pkg/config/validation/validation.go:2684]: I refacored the test case to validate **across** HTTPRoutes.

But even within an HTTPRoute, if the catchall match presents in the first one, the remaining matches are skipped as well although from data plane perspective it still works.

Comment (shonecyx) [reply pkg/config/validation/validation.go:2684]: Thanks @howardjohn. Any further comment? If no could you help merge?

**Comment by shonecyx** [comment]
@howardjohn Do you have further concerns, if not could you please approve this?

File: pkg/config/validation/validation.go
Comment (shonecyx) [reply pkg/config/validation/validation.go:2684]: ping @howardjohn

**Comment by shonecyx** [comment]
@howardjohn @hzxuzhonghu how can we move this forward?

**Comment by hzxuzhonghu** [comment]
need @howardjohn and @istio/wg-environments-maintainers to take a look

**Comment by shonecyx** [comment]
@howardjohn  any update?

**Comment by shonecyx** [comment]
ping @howardjohn

File: pkg/config/analysis/analyzers/testdata/virtualservice_overlappingmatches.yaml
Comment (hanxiaop) [inline pkg/config/analysis/analyzers/testdata/virtualservice_overlappingmatches.yaml:33]: This change seems irrelevant to me. I think without this change, the analyzer test can still pass because the validation analyzers are actually skipped in the test.

File: pilot/pkg/networking/core/v1alpha3/route/route.go
Comment (zirain) [inline pilot/pkg/networking/core/v1alpha3/route/route.go:393]: @howardjohn This looks like the missing log I was looking for when debuging https://github.com/istio/istio/pull/50109

Comment (shonecyx) [reply pilot/pkg/networking/core/v1alpha3/route/route.go:393]: I think more folks will run into this issue after upgrade istio to 1.17 or above and this log is quite helpful for troubleshooting. @howardjohn  any suggestion to move forward?

**Pull request closed without merging**
